flow/display_list.cc

```diff
@@ -0,0 +1,1248 @@
+const SkSamplingOptions DisplayList::MipmapSampling =
+    SkSamplingOptions(SkFilterMode::kLinear, SkMipmapMode::kLinear);
+const SkSamplingOptions DisplayList::CubicSampling =
+    SkSamplingOptions(SkCubicResampler{1 / 3.0f, 1 / 3.0f});
+
+enum dlCompare {
+  NOT_EQUAL,
+  BULK_COMPARE,
+  EQUAL,
+};
+
+// Assuming a 64-bit platform (most of our platforms at this time?)
+//
+// Struct allocation in the DL memory is aligned to a void* boundary
+// which means that the minimum (aligned) struct size will be 8 bytes.
+// The DLOp base uses 4 bytes so each Op-specific struct gets 4 bytes
+// of data for "free" and works best when it packs well into an 8-byte
+// aligned size.
+struct DLOp {
+  uint8_t type : 8;
+  uint32_t size : 24;
+  dlCompare equals(const DLOp* other) const { return BULK_COMPARE; }
+};
+
+// 4 byte header + 4 byte payload packs into minimum 8 bytes
+#define DEFINE_SET_BOOL_OP(name)                            \
+  struct Set##name##Op final : DLOp {                       \
+    static const auto kType = DisplayListOpType::Set##name; \
+                                                            \
+    Set##name##Op(bool value) : value(value) {}             \
+                                                            \
+    const bool value;                                       \
+                                                            \
+    void dispatch(Dispatcher& dispatcher) const {           \
+      dispatcher.set##name(value);                          \
+    }                                                       \
+  };
+DEFINE_SET_BOOL_OP(AA)
+DEFINE_SET_BOOL_OP(Dither)
+DEFINE_SET_BOOL_OP(InvertColors)
+#undef DEFINE_SET_CLEAR_BOOL_OP
+
+// 4 byte header + 4 byte payload packs into minimum 8 bytes
+#define DEFINE_SET_ENUM_OP(name)                               \
+  struct Set##name##s##Op final : DLOp {                       \
+    static const auto kType = DisplayListOpType::Set##name##s; \
+                                                               \
+    Set##name##s##Op(SkPaint::name value) : value(value) {}    \
+                                                               \
+    const SkPaint::name value;                                 \
+                                                               \
+    void dispatch(Dispatcher& dispatcher) const {              \
+      dispatcher.set##name##s(value);                          \
+    }                                                          \
+  };
+DEFINE_SET_ENUM_OP(Cap)
+DEFINE_SET_ENUM_OP(Join)
+#undef DEFINE_SET_ENUM_OP
+
+// 4 byte header + 4 byte payload packs into minimum 8 bytes
+struct SetDrawStyleOp final : DLOp {
+  static const auto kType = DisplayListOpType::SetDrawStyle;
+
+  SetDrawStyleOp(SkPaint::Style style) : style(style) {}
+
+  const SkPaint::Style style;
+
+  void dispatch(Dispatcher& dispatcher) const {
+    dispatcher.setDrawStyle(style);
+  }
+};
+// 4 byte header + 4 byte payload packs into minimum 8 bytes
```

Comment:
Instead of these comments, perhaps something like `static_assert(sizeof(SetStrokeWidthOp) == sizeof<DLOp> + ...)`. Just to make the compiler check the comment for you too.

Reply:
I don't think I can static assert since the situation may end up being more compact on 32-bit compiles - which is a good thing. These comments are about an optimization to avoid poorly packed structures wasting space unnecessarily. The values are checked in a unit test that run on 64-bit hosts, so I think that is a good compromise.

Reply:
I beefed up the comment at the top of the section that these calculations are only "worst case" numbers to double check we aren't wasting space unnecessarily.